src/opencl_cloudkeychain_fmt_plug.c

```diff
@@ -0,0 +1,434 @@
+		HANDLE_CLERROR(clReleaseMemObject(mem_out), "Release mem out");
+
+		MEM_FREE(host_pass);
+		MEM_FREE(host_salt);
+		MEM_FREE(host_crack);
+	}
+}
+
+static void done(void)
+{
+	if (autotuned) {
+		release_clobj();
+		HANDLE_CLERROR(clReleaseKernel(final_kernel), "Release kernel");
+		HANDLE_CLERROR(clReleaseKernel(split_kernel), "Release kernel");
+		HANDLE_CLERROR(clReleaseKernel(crypt_kernel), "Release kernel");
+		HANDLE_CLERROR(clReleaseProgram(program[gpu_id]), "Release Program");
+
+		autotuned--;
+	}
+}
+
+static void *get_salt(char *ciphertext)
+{
+        char *ctcopy = strdup(ciphertext);
+        char *keeptr = ctcopy;
+        int i;
+        char *p;
+        static struct custom_salt cs;
+
+        memset(&cs, 0, sizeof(cs));
+        ctcopy += FORMAT_TAG_LEN;       /* skip over "$cloudkeychain$" */
+        p = strtokm(ctcopy, "$");
+        cs.saltlen = atoi(p);
+        p = strtokm(NULL, "$");
+        for (i = 0; i < cs.saltlen; i++)
+                cs.salt[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+        p = strtokm(NULL, "$");
+        cs.iterations = atou(p);
+        p = strtokm(NULL, "$");
+        cs.masterkeylen = atoi(p);
+        p = strtokm(NULL, "$");
+        for (i = 0; i < cs.masterkeylen; i++)
+                cs.masterkey[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+        p = strtokm(NULL, "$");
+        cs.plaintextlen = atou(p);
+        p = strtokm(NULL, "$");
+        cs.ivlen = atoi(p);
+        p = strtokm(NULL, "$");
+        for (i = 0; i < cs.ivlen; i++)
+                cs.iv[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+        p = strtokm(NULL, "$");
+        cs.cryptextlen = atoi(p);
+        p = strtokm(NULL, "$");
+        for (i = 0; i < cs.cryptextlen; i++)
+                cs.cryptext[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+        p = strtokm(NULL, "$");
+        cs.expectedhmaclen = atoi(p);
+        p = strtokm(NULL, "$");
+        for (i = 0; i < cs.expectedhmaclen; i++)
+                cs.expectedhmac[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+        p = strtokm(NULL, "$");
+                cs.hmacdatalen = atoi(p);
+        p = strtokm(NULL, "$");
+        for (i = 0; i < cs.hmacdatalen; i++)
+                cs.hmacdata[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+
+	// we append the count and EOM here, one time <- salt hack required by our opencl kernel
+	memcpy(&cs.salt[cs.saltlen], "\x0\x0\x0\x1\x80", 5);
+	cs.saltlen = cs.saltlen + 5; // we include the x80 byte in our saltlen, but the .cl kernel knows to reduce saltlen by 1
+
+        MEM_FREE(keeptr);
+        return (void *)&cs;
+}
```

Comment:
Something is amiss with indentation above. Looks like spaces instead of tabs in most of the function?

Reply:
Yep, good catch. Fixed now.